platform/darwin/src/darwin_conversion.hpp

```diff
@@ -0,0 +1,132 @@
+#pragma once
+
+#include <mbgl/util/logging.hpp>
+#include <mbgl/style/conversion.hpp>
+#include <mbgl/util/feature.hpp>
+#include <mbgl/util/optional.hpp>
+
+namespace mbgl {
+namespace style {
+namespace conversion {
+
+// A minimal wrapper class conforming to the requirements for `objectMember(v, name)` (see mbgl/style/conversion.hpp)
+// This is necessary because using `NSObject*` as the value type in `optional<NSObject*>` causes problems for the ARC,
+// due to things like `optional(const value_type& __v)`
+class OptionalNSObjectValue {
+public:
+    OptionalNSObjectValue(NSObject *_value) : value(_value) {}
+    OptionalNSObjectValue() {}
+    
+    explicit operator bool() const {
+        return value;
+    }
+    
+    NSObject* operator*() {
+        assert(this);
+        return value;
+    }
+private:
+    NSObject* value;
+};
+
+inline bool isUndefined(const NSObject* value) {
+    return !value || value == [NSNull null];
+}
+
+inline bool isArray(const NSObject* value) {
+    return [value isKindOfClass:[NSArray class]];
+}
+
+inline bool isObject(const NSObject* value) {
+    return [value isKindOfClass:[NSDictionary class]];
+}
+
+inline std::size_t arrayLength(const NSObject* value) {
+    NSCAssert([value isKindOfClass:[NSArray class]], @"Value must be an NSArray for getLength().");
+    NSArray * array = (NSArray *)value;
+    return [array count];
+}
+
+inline NSObject* arrayMember(const NSObject* value, std::size_t i) {
+    NSCAssert([value isKindOfClass:[NSArray class]], @"Value must be an NSArray for get(int).");
+    NSArray * array = (NSArray *)value;
+    return array[i];
+}
+
+inline OptionalNSObjectValue objectMember(const NSObject* value, const char* key) {
+    NSCAssert([value isKindOfClass:[NSDictionary class]], @"Value must be an NSDictionary for get(string).");
+    auto dict = (NSDictionary *)value;
+    NSObject *member = dict[@(key)];
+
+    if (value && value != [NSNull null]) {
+        return { member };
+    } else {
+        return {};
+    }
+}
+
+template <class Fn>
+optional<Error> eachMember(const NSObject*, Fn&&) {
```

Comment:
Don't define this method at all. Then you'll get a compiler error if something tries to use it.